lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java

```diff
@@ -40,28 +39,40 @@
   /** The Similarity implementation. */
   final Similarity similarity;
   final BooleanQuery query;
-  
-  final ArrayList<Weight> weights;
+
+  private static class WeightedBooleanClause {
+    BooleanClause clause;
```

Comment:
I \*think\* we only need the BooleanClause#Occur here?

Reply:
The code also uses some convenience methods on `BooleanClause` (`isRequired()`, `isScoring()`, and `isProhibited()`). Those are only checking the `BooleanCause#Occur`, but these checks are used elsewhere, so we should not inline them, I suppose. Therefore, we need the entire clause.